core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/CompressRequestStage.java

```diff
@@ -0,0 +1,251 @@
+        }
+        if (input.contentStreamProvider() == null) {
+            return false;
+        }
+        return isRequestSizeWithinThreshold(input, context);
+    }
+
+    private static boolean isStreaming(RequestExecutionContext context) {
+        return context.executionAttributes().getAttribute(SdkInternalExecutionAttribute.REQUEST_COMPRESSION).isStreaming();
+    }
+
+    private void compressEntirePayload(SdkHttpFullRequest.Builder input, Compressor compressor) {
+        ContentStreamProvider wrappedProvider = input.contentStreamProvider();
+        ContentStreamProvider compressedStreamProvider = () -> compressor.compress(wrappedProvider.newStream());
+        input.contentStreamProvider(compressedStreamProvider);
+    }
+
+    private static void updateContentEncodingHeader(SdkHttpFullRequest.Builder input,
+                                                                          Compressor compressor) {
+        if (input.firstMatchingHeader("Content-encoding").isPresent()) {
+            input.appendHeader("Content-encoding", compressor.compressorType());
+        } else {
+            input.putHeader("Content-encoding", compressor.compressorType());
+        }
+    }
+
+    private static void updateContentLengthHeader(SdkHttpFullRequest.Builder input) {
+        InputStream inputStream = input.contentStreamProvider().newStream();
+        try {
+            byte[] bytes = IoUtils.toByteArray(inputStream);
+            String length = String.valueOf(bytes.length);
+            input.putHeader("Content-Length", length);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static Compressor resolveCompressorType(ExecutionAttributes executionAttributes) {
+        List<String> encodings =
+            executionAttributes.getAttribute(SdkInternalExecutionAttribute.REQUEST_COMPRESSION).getEncodings();
+
+        for (String encoding: encodings) {
+            encoding = encoding.toLowerCase(Locale.ROOT);
+            if (CompressorType.isSupported(encoding)) {
+                return CompressorType.of(encoding).newCompressor();
+            }
+        }
+        return null;
+    }
+
+    private static boolean resolveRequestCompressionEnabled(RequestExecutionContext context) {
+
+        Optional<Boolean> requestCompressionEnabledRequestLevel =
+            context.originalRequest().overrideConfiguration()
+                   .flatMap(RequestOverrideConfiguration::requestCompressionConfiguration)
+                   .map(RequestCompressionConfiguration::requestCompressionEnabled);
+        if (requestCompressionEnabledRequestLevel.isPresent()) {
+            return requestCompressionEnabledRequestLevel.get();
+        }
+
+        if (compressionEnabledClientLevel != null) {
+            return compressionEnabledClientLevel;
+        }
+        if (context.executionAttributes().getAttribute(SdkExecutionAttribute.REQUEST_COMPRESSION_CONFIGURATION) != null) {
```

Comment:
As mentioned offline, let's move the resolution logic to SdkDefaultClientBuilder to keep all client options in one place